### PR TITLE
Improving information related to APIs by adding events

### PR DESCRIPTION
**Pull request by RakhithaRR**
#### Purpose
This PR,

1. Improves the information returned by `kubectl describe apis` command by adding events related to the specific API resource. This is done by introducing an event recorder to the api-controller which will send custom events to the relevant resource.
Related issue - https://github.com/wso2/k8s-api-operator/issues/334
However, the events will be removed after 1 hour by the garbage collector in Kubernetes.

```
$ kubectl describe apis petstore-api

Events:
  Type    Reason            Age                From            Message
  ----    ------            ----               ----            -------
  Normal  Configs           15m (x3 over 18m)  api-controller  Creating swagger configmap for MGW.
  Normal  Configs           15m (x3 over 18m)  api-controller  Handling analytics & interceptors, rendering dockerfile & mgw configs, and creating the Kaniko job.
  Normal  KanikoJob         15m (x3 over 18m)  api-controller  Deploying kaniko job.
  Normal  KanikoJob         15m (x3 over 17m)  api-controller  Kaniko job completed successfully.
  Normal  MGWRuntime        15m (x3 over 17m)  api-controller  Deploying MGW runtime: petstore-api.
  Normal  MGWService        15m (x3 over 17m)  api-controller  Creating MGW service: petstore-api.
  Normal  MGWservice        15m (x2 over 15m)  api-controller  Managed API service endpoint: 35.224.76.103
  Normal  Deploy            15m (x2 over 15m)  api-controller  Successfully deployed the API: petstore-api.
```

2. Fixes an issue where reconciler is re-queued infinitely when LoadBalancer type is used with Minikube clusters.

3. Changes the controller logic to create concurrent reconcilers.

#### Suggested Labels
Improvement

### CODE REVIEW

File: api-operator/pkg/controller/api/api_controller.go [hunk 1]
```diff
@@ -480,36 +517,63 @@ func (r *ReconcileAPI) Reconcile(request reconcile.Request) (reconcile.Result, e
 			return reconcile.Result{}, errHpa
 		}
 
+		for t := 24; t > 0; t -= 1 {
+			time.Sleep(5 * time.Second)
+			getEndPointValue := mgw.ExternalIP(&r.client, instance, operatorMode, mgwSvc, controlIngressData, controlOpenshiftConf)
```
Comment (renuka-fernando) [inline api-operator/pkg/controller/api/api_controller.go:522]: we have to update the value of `mgwSvc` by calling `k8s.Get()`, and then call the function `mgw.ExternalIP()`

Comment (RakhithaRR) [reply api-operator/pkg/controller/api/api_controller.go:525]: fixed in dc9a0dc

File: api-operator/pkg/controller/api/api_controller.go [hunk 2]
```diff
@@ -365,23 +386,29 @@ func (r *ReconcileAPI) Reconcile(request reconcile.Request) (reconcile.Result, e
...
+				r.recorder.Event(instance, eventTypeError, "Configs", "Error while handling interceptors.")
 				return reconcile.Result{}, err
 			}
 
 			// handling Kaniko docker file
 			log.Info("Rendering the dockerfile for Kaniko job and adding volumes to the Kaniko job")
 			if err := kaniko.HandleDockerFile(&r.client, userNamespace, instance.Name, ownerRef); err != nil {
 				log.Error(err, "Error rendering the docker file for Kaniko job and adding volumes to the Kaniko job")
+				r.recorder.Event(instance, eventTypeError, "KanikoJob",
+					"Error rendering the dockerfile for kaniko job.")
```
Comment (kavishkafernando) [inline api-operator/pkg/controller/api/api_controller.go:411]: Shouldn't the k be capitalized in 'Error rendering the dockerfile for **kaniko** job'

File: api-operator/pkg/controller/api/api_controller.go [hunk 3]
```diff
@@ -407,12 +434,15 @@ func (r *ReconcileAPI) Reconcile(request reconcile.Request) (reconcile.Result, e
...
 			log.Info("Deploying the Kaniko job in cluster")
+			r.recorder.Event(instance, corev1.EventTypeNormal, "KanikoJob", "Deploying kaniko job.")
 			kanikoJob = kaniko.Job(instance, controlConfigData, kanikoArgs.Data[kanikoArguments], ownerRef)
 			if err := controllerutil.SetControllerReference(instance, kanikoJob, r.scheme); err != nil {
 				return reconcile.Result{}, err
 			}
 			// create Kaniko job and set kaniko object
 			if errJob := k8s.CreateIfNotExists(&r.client, kanikoJob); errJob != nil {
+				r.recorder.Event(instance, eventTypeError, "KanikoJob",
+					"Error when creating the kaniko job.")
```
Comment (kavishkafernando) [inline api-operator/pkg/controller/api/api_controller.go:445]: same here

File: api-operator/pkg/controller/api/api_controller.go (hunk 2 above)
Comment (RakhithaRR) [reply api-operator/pkg/controller/api/api_controller.go:411]: Those have been removed in the latest code.

File: api-operator/pkg/controller/api/api_controller.go (hunk 3 above)
Comment (RakhithaRR) [reply api-operator/pkg/controller/api/api_controller.go:445]: Those have been removed in the latest code.

**Pull request merged**